internal/app/machined/pkg/controllers/secrets/trustd.go

```diff
@@ -0,0 +1,273 @@
+				continue
+			}
+
+			return fmt.Errorf("error getting machine type: %w", err)
+		}
+
+		machineType := machineTypeRes.MachineType()
+
+		if !machineType.IsControlPlane() {
+			return fmt.Errorf("machine type changed")
+		}
+
+		rootResource, err := safe.ReaderGet[*secrets.OSRoot](ctx, r, resource.NewMetadata(secrets.NamespaceName, secrets.OSRootType, secrets.OSRootID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				if err = ctrl.teardownAll(ctx, r); err != nil {
+					return fmt.Errorf("error destroying resources: %w", err)
+				}
+
+				continue
+			}
+
+			return fmt.Errorf("error getting etcd root secrets: %w", err)
+		}
+
+		rootSpec := rootResource.TypedSpec()
+
+		certSANResource, err := safe.ReaderGet[*secrets.CertSAN](ctx, r, resource.NewMetadata(secrets.NamespaceName, secrets.CertSANType, secrets.CertSANAPIID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting certSANs: %w", err)
+		}
+
+		certSANs := certSANResource.TypedSpec()
+
+		if err := ctrl.generateControlPlane(ctx, r, logger, rootSpec, certSANs); err != nil {
+			return err
+		}
+	}
+}
+
+func (ctrl *TrustdController) generateControlPlane(ctx context.Context, r controller.Runtime, logger *zap.Logger, rootSpec *secrets.OSRootSpec, certSANs *secrets.CertSANSpec) error {
+	ca, err := x509.NewCertificateAuthorityFromCertificateAndKey(rootSpec.CA)
+	if err != nil {
+		return fmt.Errorf("failed to parse CA certificate: %w", err)
+	}
+
+	serverCert, err := x509.NewKeyPair(ca,
+		x509.IPAddresses(certSANs.StdIPs()),
+		x509.DNSNames(certSANs.DNSNames),
+		x509.CommonName(certSANs.FQDN),
+		x509.NotAfter(time.Now().Add(x509.DefaultCertificateValidityDuration)),
+		x509.KeyUsage(stdlibx509.KeyUsageDigitalSignature|stdlibx509.KeyUsageKeyEncipherment),
+		x509.ExtKeyUsage([]stdlibx509.ExtKeyUsage{
+			stdlibx509.ExtKeyUsageServerAuth,
+		}),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to generate API server cert: %w", err)
+	}
+
+	if err := safe.WriterModify(ctx, r, secrets.NewTrustd(),
+		func(r *secrets.Trustd) error {
+			trustdSecrets := r.TypedSpec()
+
+			trustdSecrets.CA = &x509.PEMEncodedCertificateAndKey{
+				Crt: rootSpec.CA.Crt,
+			}
+			trustdSecrets.Server = x509.NewCertificateAndKeyFromKeyPair(serverCert)
+
+			return nil
+		}); err != nil {
+		return fmt.Errorf("error modifying resource: %w", err)
+	}
+
+	serverFingerprint, _ := x509.SPKIFingerprintFromDER(serverCert.Certificate.Certificate[0]) //nolint:errcheck
+
+	logger.Debug("generated new certificates",
+		zap.Stringer("server", serverFingerprint),
+	)
+
+	return nil
+}
+
+func (ctrl *TrustdController) teardownAll(ctx context.Context, r controller.Runtime) error {
+	list, err := r.List(ctx, resource.NewMetadata(secrets.NamespaceName, secrets.TrustdType, "", resource.VersionUndefined))
+	if err != nil {
+		return err
+	}
+
+	// TODO: change this to proper teardown sequence
```

Comment:
this teardown seems fine to me :sweat_smile: , anything else extra needed here?

Reply:
yep, probably I should remove it, it's an old comment I got via copy-paste